wagtail_newsletter/models.py

```diff
@@ -142,6 +143,10 @@ def save_revision(self, *args, **kwargs):
         )
         return revision
 
+    def has_newsletter_permission(self, user, action):
+        permission_policy = ModelPermissionPolicy(type(self))
+        return permission_policy.user_has_permission(user, "publish")
```

Comment:
Just to be sure I understand correctly what we're doing here: the reason why these permissions are _not_ declared by default on the NewsletterPageMixin's Meta, is that we want models that use this mixin to explicitly declare the subset of these permissions they need?

If that's the case, I wonder if the Article model of that "demo" app shouldn't declare such permissions in its Meta, so they would show up in the Wagtail UI that is in charge of assigning permissions to groups?

Reply:
> Just to be sure I understand correctly what we're doing here: the reason why these permissions are _not_ declared by default on the NewsletterPageMixin's Meta, is that we want models that use this mixin to explicitly declare the subset of these permissions they need?

Yes. And they should subclass `has_newsletter_permission` to enforce the permissions.

Also because I didn't know enough about Django permissions and declaring them on Meta 🙈 But, from what I can tell, the permission will be bound to the model's app, therefore `wagtail-newsletter` doesn't know its full name. Also, the permission typically includes the model's name, which is not known to the mixin. Wagtail creates a [dummy model](https://github.com/wagtail/wagtail/blob/4e2afc966bc4b37dd425d99c83cfe715563ffcc8/wagtail/admin/models.py#L15-L23) in order to define a global permission which feels too heavy-handed here.

> If that's the case, I wonder if the Article model of that "demo" app shouldn't declare such permissions in its Meta, so they would show up in the Wagtail UI that is in charge of assigning permissions to groups?

That's a great idea! I've added a few things to the demo app:
- Declare a `sendnewsletter_articlepage` permission.
- Register permissions for the custom recipients model, and the `sendnewsletter_articlepage` permission, so they show up in Wagtail's permissions editor.
- Hide the newsletter tab entirely (by hiding all of its panels) if the user doesn't have the `sendnewsletter_articlepage` permission.

I've also changed `NewsletterPanel` to be hidden entirely if the user doesn't have permission for any of its actions.

I wish I had these improvements in place for the demo yesterday :)